File: frontend/lib/shared/services/notification_service.dart
import '../../core/constants/api_constants.dart';
import '../../core/services/cache_manager.dart';
import 'api_client.dart';

class NotificationService {
  final ApiClient _api = ApiClient.instance;
  final CacheManager _cache = CacheManager.instance;

  /// Get notifications for a coaching — SWR stream.
  Stream<Map<String, dynamic>> watchCoachingNotifications(
    String coachingId, {
    int limit = 20,
    int offset = 0,
  }) {
    final key = 'notif:coaching:$coachingId:$limit:$offset';
    return _cache.swr<Map<String, dynamic>>(
      key,
      () => _api.getAuthenticated(
        '${ApiConstants.coaching}/$coachingId/notifications?limit=$limit&offset=$offset',
      ),
      (raw) => Map<String, dynamic>.from(raw as Map),
    );
  }

  Future<Map<String, dynamic>> getCoachingNotifications(
    String coachingId, {
    int limit = 20,
    int offset = 0,
  }) =>
      watchCoachingNotifications(coachingId, limit: limit, offset: offset).last;

  /// Get personal notifications — SWR stream.
  Stream<Map<String, dynamic>> watchUserNotifications({
    int limit = 20,
    int offset = 0,
  }) {
    final key = 'notif:user:$limit:$offset';
    return _cache.swr<Map<String, dynamic>>(
      key,
      () => _api.getAuthenticated(
        '${ApiConstants.baseUrl}/notifications/me?limit=$limit&offset=$offset',
      ),
      (raw) => Map<String, dynamic>.from(raw as Map),
    );
  }

  Future<Map<String, dynamic>> getUserNotifications({
    int limit = 20,
    int offset = 0,
  }) => watchUserNotifications(limit: limit, offset: offset).last;

  /// Mark a notification as read.
  Future<void> markAsRead(String notificationId) async {
    await _api.patchAuthenticated(
      '${ApiConstants.baseUrl}/notifications/$notificationId/read',
      body: {},
    );
    await _cache.invalidatePrefix('notif:');
  }

  /// Archive a notification (mark as removed from view without deleting).
  Future<void> archiveNotification(String notificationId) async {
    await _api.patchAuthenticated(
      '${ApiConstants.baseUrl}/notifications/$notificationId/archive',
      body: {},
    );
    await _cache.invalidatePrefix('notif:');
  }

  /// Delete a notification.
  Future<void> deleteNotification(String notificationId) async {
    await _api.deleteAuthenticated(
      '${ApiConstants.baseUrl}/notifications/$notificationId',
    );
    await _cache.invalidatePrefix('notif:');
  }
}

File: backend/src/modules/notification/notification.service.ts
import prisma from '../../infra/prisma.js';

export interface CreateNotificationDto {
    coachingId?: string;
    userId?: string;
    type: string;
    title: string;
    message: string;
    data?: any;
}

export class NotificationService {
    /**
     * Get notifications for a coaching (admin view).
     * Excludes archived notifications.
     */
    async getCoachingNotifications(coachingId: string, limit = 20, offset = 0) {
        return prisma.notification.findMany({
            where: { coachingId, archived: false },
            orderBy: { createdAt: 'desc' },
            take: limit,
            skip: offset,
        });
    }

    /**
     * Get personal notifications for a user.
     * Excludes archived notifications.
     */
    async getUserNotifications(userId: string, limit = 20, offset = 0) {
        return prisma.notification.findMany({
            where: { userId, archived: false },
            orderBy: { createdAt: 'desc' },
            take: limit,
            skip: offset,
        });
    }

    /**
     * Get Unread count for coaching (excludes archived)
     */
    async getCoachingUnreadCount(coachingId: string) {
        return prisma.notification.count({
            where: { coachingId, read: false, archived: false },
        });
    }

    /**
     * Get Unread count for user (excludes archived)
     */
    async getUserUnreadCount(userId: string) {
        return prisma.notification.count({
            where: { userId, read: false, archived: false },
        });
    }

    /**
     * Mark a notification as read.
     */
    async markAsRead(notificationId: string) {
        return prisma.notification.update({
            where: { id: notificationId },
            data: { read: true },
        });
    }

    /**
     * Delete a notification.
     */
    async deleteNotification(notificationId: string) {
        return prisma.notification.delete({
            where: { id: notificationId },
        });
    }

    /**
     * Archive a notification (mark as archived instead of deleting).
     */
    async archiveNotification(notificationId: string) {
        return prisma.notification.update({
            where: { id: notificationId },
            data: { archived: true },
        });
    }

    /**
     * Create a notification (internal use).
     */
    async create(data: CreateNotificationDto) {
        return prisma.notification.create({
            data: {
                ...(data.coachingId && { coachingId: data.coachingId }),
                ...(data.userId && { userId: data.userId }),
                type: data.type,
                title: data.title,
                message: data.message,
                data: data.data,
            },
        });
    }
}
